src/reconcile/rec_visibility.c

```diff
@@ -252,6 +257,112 @@ __timestamp_out_of_order_fix(WT_SESSION_IMPL *session, WT_TIME_WINDOW *select_tw
     return (false);
 }
 
+/*
+ * __rec_validate_upd_chain --
+ *     Check the update chain for conditions that would prevent its insertion into the history
+ *     store. Return EBUSY if the update chain cannot be inserted into the history store at this
+ *     time.
+ */
+static int
+__rec_validate_upd_chain(WT_SESSION_IMPL *session, WT_RECONCILE *r, WT_UPDATE *select_upd,
+  WT_TIME_WINDOW *select_tw, WT_CELL_UNPACK_KV *vpack)
+{
+    WT_DECL_RET;
+    WT_UPDATE *prev_upd, *upd;
+
+    /*
+     * There is no selected update to go to disk as such we don't need to check the updates
+     * following it.
+     */
+    if (select_upd == NULL)
+        return (0);
+
+    /* There is no history store so it is safe to reconcile out of order updates. */
+    if (!F_ISSET(S2C(session), WT_CONN_HS_OPEN))
+        return (0);
+
+    /* History store out of order update is allowed. */
+    if (WT_IS_HS(session->dhandle))
+        return (0);
+
+    /*
+     * If we're not in eviction any history store insertion is fine, if we are in eviction and a
+     * checkpoint isn't running then history store insertions are also fine.
+     */
+    if (!F_ISSET(r, WT_REC_CHECKPOINT_RUNNING))
+        return (0);
+
+    /*
+     * The selected time window may contain information that isn't visible given the selected
+     * update, as such we have to check it separately. This is true when there is a tombstone ahead
+     * of the selected update.
+     */
+    if (select_tw->stop_ts < select_tw->start_ts)
+        WT_ERR(EBUSY);
+
+    /* The selected update is restored from the data store or history store. */
```

Comment:
Does that make it OK because any out-of-order updates would have been resolved when it was written last time? If so a more explicit comment would help.

Reply:
Resolving as comment was improved.